devfile/devworkspace-operator
# feat: Allow external DWOC to be merged with internal DWOC on a per-workspace basis

### What does this PR do?
This PR aims to make 2 changes to DWO:
1. Refactor (almost) all uses of the DWOC, so that, internally, each workspace instance has its own configuration. Currently, this per-workspace devworkspace-operator configuration is not visible to users, and was added only to support changed n.2
2. Allow for specifying an external DWOC (that exists on the cluster), which gets merged with the workspace's internal DWOC

The external DWOC's name and namespace are specified with the `controller.devfile.io/devworkspace-config` DevWorkspace attribute, which has the following structure:
    
```YAML
attributes:
  controller.devfile.io/devworkspace-config:
    name: <string>
    namespace: <string>
```


### What issues does this PR fix or reference?
Part of eclipse/che#21405

### Is it tested? How?
I added 2 automated tests.

For manual testing, perform the following steps:
1. Deploy DWO
2. Create a devfile with the `controller.devfile.io/devworkspace-config` attribute set, e.g.

```YAML
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: theia-next-external-dwoc
spec:
  started: true
  template:
    attributes:
        controller.devfile.io/devworkspace-config:
          name: external-dwoc-test
          namespace: default
        controller.devfile.io/storage-type: per-workspace
    projects:
      - name: web-nodejs-sample
        git:
          remotes:
            origin: "https://github.com/che-samples/web-nodejs-sample.git"
    components:
      - name: theia
        plugin:
          uri: https://che-plugin-registry-main.surge.sh/v3/plugins/eclipse/che-theia/next/devfile.yaml
          components:
            - name: theia-ide
              container:
                env:
                  - name: THEIA_HOST
                    value: 0.0.0.0
    commands:
      - id: say-hello
        exec:
          component: theia-ide
          commandLine: echo "Hello from $(pwd)"
          workingDir: ${PROJECTS_ROOT}/project/app
```
3. Create a DWOC file, for reference, we'll call it `test-external-dwoc.yaml`. Also, ensure its name field is set to the same value as the `controller.devfile.io/devworkspace-config-name` attribute in the devfile from step 2:
```YAML
apiVersion: controller.devfile.io/v1alpha1
config:
  enableExperimentalFeatures: true
  routing:
    clusterHostSuffix: 192.168.49.2.nip.io
    defaultRoutingClass: basic
  workspace:
    defaultStorageSize:
      common: 20Gi
      perWorkspace: 29Gi
    imagePullPolicy: Always
kind: DevWorkspaceOperatorConfig
metadata:
  name: external-dwoc-test
  namespace: default
```
4. Apply the (external) DWOC file you just created, ensuring it's in the namespace you specified in the `controller.devfile.io/devworkspace-config-namespace` (in my example, the default namespace is used): `kubectl apply -f test-external-dwoc.yaml`
5. Apply the devfile created in step 2. 
6. Verify that the internal DWOC has been modified. I recommend checking the size of the per-workspace PVC that is created and ensuring it is `29Gi` instead of the default `10Gi`
7. Afterwards, you can ensure that workspace's that do not have the `controller.devfile.io/devworkspace-config` attribute set are not affected by the external DWOC, and instead, they use the global/internal DWOC. For instance, start up the per-workspace example and ensure the PVC created for it is `10Gi` big: `kubectl apply -f per-workspace-storage.yaml -n $NAMESPACE`.

### Some notes and limitations:
1. The HTTP client / Proxy settings cannot be affected by the externally applied DWOC, as [the HTTP client is created and configured on DWO startup rather than workspace startup](https://github.com/devfile/devworkspace-operator/pull/711#issuecomment-995360107). 
2. There is [a handler](https://github.com/devfile/devworkspace-operator/blob/main/controllers/workspace/devworkspace_controller.go#L680) that checks if the common PVC was deleted. Currently, it filters for PVC's related to DWO by checking the PVC name against the global configs PVC name. In the future, this limitation can be worked around by applying a DWO-specific annotation to all PVC's created by DWO, and filtering by that annotation. 
3. This change could have weird implications for anything configurable that is shared across workspaces. Most notably, the common PVC. For example, if the common PVC does not yet exist on the cluster, and an external DWOC that has the common PVC size set to `20Gi` is applied for a specific workspace, then other (future) workspaces that use the common storage strategy will also be using a PVC of `20Gi`, even if they aren't using an external DWOC. 

### PR Checklist
